cgp/ea/mu_plus_lambda.py

```diff
@@ -137,13 +137,34 @@ def step(
         combined = self._compute_fitness(combined, objective)
         combined = self._sort(combined)
 
-        n_total = self.n_offsprings + pop.n_parents
-        k_local_search = n_total if self.k_local_search is None else self.k_local_search
-        for idx in range(k_local_search):
-            self.local_search(combined[idx])
+        if self.local_search is not None:
+            assert isinstance(pop.champion.fitness, float)
+            prev_max_fitness: float = pop.champion.fitness
 
-        combined = self._compute_fitness(combined, objective)
-        combined = self._sort(combined)
+            combined_copy = [ind.copy() for ind in combined]
+
+            k_local_search = (
+                len(combined_copy) if self.k_local_search is None else self.k_local_search
+            )
+            for idx in range(k_local_search):
+                self.local_search(combined_copy[idx])
+
+            combined_copy = self._compute_fitness(combined_copy, objective)
+
+            new_combined = []
+            for ind in combined:
+                for ind_copy in combined_copy:
+                    if ind.idx == ind_copy.idx:
+                        assert ind.fitness is not None
+                        assert ind_copy.fitness is not None
+                        if ind.fitness < ind_copy.fitness:
+                            new_combined.append(ind_copy)
+                        else:
+                            new_combined.append(ind)
+
+            combined = self._sort(new_combined)
+
+            assert pop.champion.fitness >= prev_max_fitness
```

Comment:
why is this necessary?

Reply:
strictly speaking it's not necessary, it's just a consistency check that the local search did not decrease the maximal fitness. we could also check whether it didn't decrease the average fitness of the population, maybe that's better?

Reply:
My problem with this is "what if this happens?". The search is aborted but what should the user do then? Change some parameters? Abandon this method altogether?

Has this happened in your tests? If yes, do you have an idea why this happened? At the very least, we should add a meaningful error message, ideally with some sort of hint of what the user can do.

Reply:
this should never happen unless there is an programming error somewhere in the double loop right above. error messages hence do not make sense, this is rather a protection against mistakes there, as well as documentation ("the maximal fitness should be larger or equal than before"). this has happened a lot before, but those problems have been fixed ;)

Reply:
So, if this has been fixed, we can remove the assert statement, right? Is this like a parachute because you're not 100% sure that the code is correct?

Reply:
i'm sure the code is correct, for now ;) it's a parachute for future generations if you will. the script should exit if this condition is not fulfilled for whatever as something terrible went wrong.

Reply:
Okay. I would either remove it or add an informative error message like
```The fitness decreased after executing the local search. This indicates that something went wrong during the optimization. Therefore aborting.```

I would prefer to just remove it.

Reply:
Imo, the message would be fine too

Reply:
ok, replaced with proper error message